[VectorCombine] Remove dead node immediately in VectorCombine

davemgreen opened the pull request:
The vector combiner will process all instructions as it first loops through the function, adding any newly added and deleted instructions to a worklist which is then processed when all nodes are done. These leaves extra uses in the graph as the initial processing is performed, leading to sub-optimal decisions being made for other combines. This changes it so that trivially dead instructions are removed immediately. The main change that this requires is to make sure iterator invalidation does not occur.

nikic reviewed (commented):
Compile-time: https://llvm-compile-time-tracker.com/compare.php?from=60579ec3059b2b6cc9dad90eaac1ed363fc395a7&to=7322d4b393c94af7164439419100d9c25df1e420&stat=instructions:u

Adding the entire function to the worklist at once is quite expensive...

Possibly we could do something more targeted at removing dead instructions early?

davemgreen commented:
> Compile-time: https://llvm-compile-time-tracker.com/compare.php?from=60579ec3059b2b6cc9dad90eaac1ed363fc395a7&to=7322d4b393c94af7164439419100d9c25df1e420&stat=instructions:u
> 
> Adding the entire function to the worklist at once is quite expensive...
> 
> Possibly we could do something more targeted at removing dead instructions early?

Yeah it is probably not be very useful for a lot of node - many will not be vector instructions. I will look into erasing the instructions as we replace them if we can.

davemgreen commented:
I think this is OK to go now. It was running into issues, but those have been ironed out. I moved one unrelated part into https://github.com/llvm/llvm-project/pull/152675.

RKSimon reviewed (approved):
LGTM - one trivial - but please update against trunk latest before merging in case there's any recent changes

davemgreen committed:
[VectorCombine] Remove dead node immediately in VectorCombine

The vector combiner will process all instructions as it first loops through the
function, adding any newly added and deleted instructions to a worklist which
is then processed when all nodes are done. These leaves extra uses in the graph
as the initial processing is performed, leading to sub-optimal decisions being
made for other combines. This changes it so that trivially dead instructions
are removed immediately. The main change that this requires is to make sure
iterator invalidation does not occur.
